### PR TITLE
Inout fix

### PR DESCRIPTION
**Pull request by Dolu1990**
This PR reimplement how Analog are managed in order to allows much more flexibility and also fix existing issues.

Now, analog connections are tracked bit by bit (Island) then aggreged together for the final connection.
All the assignements between analogs are deleted and then regenerated into a solved manner.

This is kind of a big change XD

### CODE REVIEW

**Review by andreasWallner — CHANGES_REQUESTED** [review]
First off: thanks for spending the effort on this! 

Some good news first: I had a look at my initial issue - that is fixed and Vivado is also happy with the outcome.

I had a bit of a look on the code and noticed a few corner cases

In addition we have a small inconsistency that I did not want to hunt down today (too late ;-)):
We can leave all analog inputs of one signal unconnected
```scala
  case class Top() extends Component {
    val y = inout(Analog(Bits(2 bit)))
  }
```
But once we connect one of them, we have to all?
```scala
  case class Top() extends Component {
    val y = inout(Analog(Bits(2 bit)))
    y(0) := True
  }
```

File: core/src/main/scala/spinal/core/internals/Phase.scala [hunk 1]
```diff
@@ -286,158 +287,249 @@ class PhaseApplyIoDefault(pc: PhaseContext) extends PhaseNetlist{
...
+              }
+            }
+          }
         }
-      })
-      //      }
-    })
+        for(bitId <- 0 until widthOf(seed)){
+          val flushes = ArrayBuffer[(AggregateKey, AggregateCtx)]()
+          val island = bitToIsland(Bit(seed, bitId, c.dslBody))
```
[the remark this answers is not in the document]
Comment (andreasWallner) [reply core/src/main/scala/spinal/core/internals/Phase.scala:513]: w.r.t. the second "issue": forget that - that was because of a local change I during testing

**Comment by Dolu1990** [comment]
> First off: thanks for spending the effort on this!
> 
> Some good news first: I had a look at my initial issue - that is fixed and Vivado is also happy with the outcome.
> 
> I had a bit of a look on the code and noticed a few corner cases
> 
> In addition we have a small inconsistency that I did not want to hunt down today (too late ;-)): We can leave all analog inputs of one signal unconnected
> 
> ```scala
>   case class Top() extends Component {
>     val y = inout(Analog(Bits(2 bit)))
>   }
> ```
> 
> But once we connect one of them, we have to all?
> 
> ```scala
>   case class Top() extends Component {
>     val y = inout(Analog(Bits(2 bit)))
>     y(0) := True
>   }
> ```

Fixed with https://github.com/SpinalHDL/SpinalHDL/pull/1046/commits/28636e82af0f8062e74c6f5683d9642a8e3b559d

File: core/src/main/scala/spinal/core/internals/Phase.scala [hunk 2]
```diff
@@ -286,158 +287,249 @@ class PhaseApplyIoDefault(pc: PhaseContext) extends PhaseNetlist{
...
+    pc.walkComponents { c =>
+      val islands = mutable.LinkedHashSet[Island]()
+      val bitToIsland = mutable.HashMap[Bit, Island]()
+
+      // Generate the list of islands
+      c.dslBody.walkStatements {
+        case s: AssignmentStatement => {
+          val targetBt = s.finalTarget
+          val sourceBt : BaseType = s.source match {
+            case bt: BaseType => bt
```
[the remark this answers is not in the document]
Comment (Dolu1990) [reply core/src/main/scala/spinal/core/internals/Phase.scala:317]: So, there is first something else.

myAnalogBits(0) will return a non analog Bool()
So 
myAnalogBool := myAnalogBits(0) 

Will drive myAnalogBool with the readed value of myAnalogBits(0) , It will not do a analog connection.

myAnalogBool := myAnalogBits(0).setAsAnalog() will do a analog connection

So yeah, to resume the current state : 
```scala
myAnalogBits(range) := ??? will do a analog connection
??? := myAnalogBits(range) will read the analog value of myAnalogBits and use it to drive ???
??? := myAnalogBits(range).setAsAnalog()) will do a analog connection
```

So there is some logic behind it.
Not sure actualy if that's what we want.

Then when this is established, BitsBitAccessFloating should not be allowed on analog i think

[the remark this answers is not in the document]
Comment (Dolu1990) [reply core/src/main/scala/spinal/core/internals/Phase.scala:318]: I pushed an update to have : 

```scala
myAnalogBits(range) := ??? will do a analog connection
??? := myAnalogBits(range) will do a analog connection
??? := myAnalogBits(range).setAsComb() will read myAnalogBits(range). to drive ???
```

I think that's less surprising like this.

File: core/src/main/scala/spinal/core/internals/Phase.scala [hunk 3]
```diff
@@ -286,158 +287,249 @@ class PhaseApplyIoDefault(pc: PhaseContext) extends PhaseNetlist{
...
+              case bt: BaseType => (0 until bt.getBitsWidth)
+              case e: BitAssignmentFixed => (e.bitId to e.bitId)
+              case e: RangedAssignmentFixed => (e.lo to e.hi)
+            }
+            val sourceRange = s.source match {
+              case bt: BaseType => (0 until bt.getBitsWidth)
+              case e: BitVectorBitAccessFixed => (e.bitId to e.bitId)
+              case e: BitVectorRangedAccessFixed => (e.lo to e.hi)
+            }
+            assert(targetRange.size == sourceRange.size)
```
[the remark this answers is not in the document]
Comment (Dolu1990) [reply core/src/main/scala/spinal/core/internals/Phase.scala]: added <3

[the remark this answers is not in the document]
Comment (Dolu1990) [reply core/src/main/scala/spinal/core/internals/Phase.scala]: Assignement overlap in the case of analog connection is ok.

File: core/src/main/scala/spinal/core/internals/Phase.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (Dolu1990) [reply core/src/main/scala/spinal/core/internals/Phase.scala]: The issue you had was that sub.x := y(0 downto 0).setAsAnalog() was required when you tested it.

Now that .setAsAnalog()  isn't necessary anymore (today push) it is all good.

Can you confirm ? (as far as i can see)

**Comment by andreasWallner** [comment]
Most of the issues are gone, but the changes now seem to have broken the "normal" generation of a driver for the `inout` (but only if the pad signal is multi-bit...):
```scala
    val x = new Component {
      val read = out Bool()
      val write = in Bool()
      val en = in Bool()
      val pad = inout(Analog(Bits(width bit)))

      for (i <- 0 until width) {
        read := pad(i)
        when(en) {
          pad(i) := write
        }
      }
    }
```
in which case the current code suspects that `MULTIPLE INOUT interconnected in the same component`

This time I pushed my test cases (currently w/o real verification) here: https://github.com/andreasWallner/SpinalHDL/tree/inout-fix

**Comment by Dolu1990** [comment]
> Most of the issues are gone, but the changes now seem to have broken the "normal" generation of a driver for the `inout` (but only if the pad signal is multi-bit...):
> 
> ```scala
>     val x = new Component {
>       val read = out Bool()
>       val write = in Bool()
>       val en = in Bool()
>       val pad = inout(Analog(Bits(width bit)))
> 
>       for (i <- 0 until width) {
>         read := pad(i)
>         when(en) {
>           pad(i) := write
>         }
>       }
>     }
> ```
> 
> in which case the current code suspects that `MULTIPLE INOUT interconnected in the same component`
> 
> This time I pushed my test cases (currently w/o real verification) here: https://github.com/andreasWallner/SpinalHDL/tree/inout-fix


Should be fixed with https://github.com/SpinalHDL/SpinalHDL/commit/56b956c9587a943655f02fd9ccb0fb12a118d3c2



Note : 

```scala
    val read = out Bits(width bit) //!!!
    val write = in Bool()
    val en = in Bool()
    val pad = inout(Analog(Bits(width bit)))

    for (i <- 0 until width) {
      read(i) := pad(i)
      when(en) {
        pad(i) := write
      }
    }
```

**Review by andreasWallner — APPROVED** [review]
Thanks for all the effort!

LGTM - should we also pull add a few testcases
I added a few here: https://github.com/andreasWallner/SpinalHDL/tree/inout-fix - while there is currently no simulation with them, but at least they are testing generation

**Comment by Dolu1990** [comment]
> should we also pull add a few testcases

It would be great, especialy corner cases.

It was my plan to add some, but currently i'm in a 2 week work trip, so realy busy, i realy lagging behind, trying to catch up emails XD

**Comment by andreasWallner** [comment]
I'm so sorry, writing tests it looks like I stumbled over another corner case:
```scala
object Blub extends App {
  SpinalVerilog(new Component {
    val pad = inout(Analog(Bits(2 bit)))
    val sliced_sub = new Component {
      val pad = inout(Analog(Bits(1 bit)))
    }
    pad(0 downto 0) <> sliced_sub.pad
  })
}
```
Which fails with
```
Static bits extraction (1 downto 0) is outside the range (0 downto 0) of (toplevel/sliced_sub/pad : inout Bits[1 bits]) at
    spinal.tester.scalatest.Blub$.delayedEndpoint$spinal$tester$scalatest$Blub$1(AnalogConnectionTester.scala:52)
```

Also: something seems to be weird with `apply` since `pad.apply(0 downto 1) <> sliced_sub.pad` is not cought early but trips over the dummy error in `Phase.scala@326`

**Comment by Dolu1990** [comment]
> I'm so sorry, writing tests it looks like I stumbled over another corner case:
> 
> ```scala
> object Blub extends App {
>   SpinalVerilog(new Component {
>     val pad = inout(Analog(Bits(2 bit)))
>     val sliced_sub = new Component {
>       val pad = inout(Analog(Bits(1 bit)))
>     }
>     pad(0 downto 0) <> sliced_sub.pad
>   })
> }
> ```
> 
> Which fails with
> 
> ```
> Static bits extraction (1 downto 0) is outside the range (0 downto 0) of (toplevel/sliced_sub/pad : inout Bits[1 bits]) at
>     spinal.tester.scalatest.Blub$.delayedEndpoint$spinal$tester$scalatest$Blub$1(AnalogConnectionTester.scala:52)
> ```
> 
> Also: something seems to be weird with `apply` since `pad.apply(0 downto 1) <> sliced_sub.pad` is not cought early but trips over the dummy error in `Phase.scala@326`

Good catch ^^
Thanks ! I will take a look.

**Comment by Dolu1990** [comment]
@andreasWallner Hi ^^
I pushed a fix for the two last issue you mentionned.

**Comment by andreasWallner** [comment]
@Dolu1990 
Thanks!
I pushed my testcases - it would be quite interesting whether they work on e.g. gitlab CI, on my machine the VHDL stuff is doing weird things (i.e. a write & writeEnable are high and the pad does not get driven to 1 after a delay)
Verilog passes.

I initially tried to also test the direction from driving the `pad`, having `writeEnable` switched off and checking `read` - but there cocotb seems to do weird stuff (when trying to deposit/release on the inout signal). I left that testcase in for reference but I could not get it to do something useful, e.g. what cocotb reports as signal levels does not match what ends up in the vcd...

**Comment by Dolu1990** [comment]
@andreasWallner I pushed a thing to disable VHDL from AnalogConnectionCocotbBoot.
So, the VHDL fail on my machine too.
You think it's because of GHDL/Cocotb issue ? or SpinalHDL VHDL generation issue ?

**Comment by andreasWallner** [comment]
I hat a look at the generated VHDL and that looks good to me... just to be sure I'll try testing it with xcelium in case I overlooked something, I should find some time to do that today.

**Comment by andreasWallner** [comment]
@Dolu1990 I had a look - it looks like the issue with the original test cases was that writing the pad signals seems to work like a force, overriding the signal fully, `Release` also did not help there. I now just removed all the writes to `xxx_pad` - now they pass...

**Comment by Dolu1990** [comment]
Nice :D
So i think all is ready to merge right ?

**Comment by andreasWallner** [comment]
@Dolu1990 sounds good to me!
